### DISCUSSION TITLE
router.push() waits for server action to resolve before navigation

### DISCUSSION THREAD

**Discussion by lukepearce** [vercel/next.js#56208] [Help]
### Summary

I’m using a server action on submit of a form in a client component. When I submit the form, I want to navigate to a page, to display a loading state while waiting for the response from the server action. I have `router.push(‘route’)` as the first line in my `onSubmit` function... but the navigation only takes place once the server action promise is resolved. I have tested this by adding a timeout to the server action.

Is this the expected behaviour? Is there a workaround?

```actions.ts
"use server";

export const addToCart = async () => {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve({
        success: true,
      });
    }, 10000);
  });
};
```

```component.tsx
"use client"

export default function Product() {
  // This version navigates instantly
  async function handleAddToCart(e) {
    e.preventDefault();
    router.push(`/cart`);
  }
  
  // This version navigates only once addToCart resolves (after 10 seconds)
  async function handleAddToCart(e) {
    e.preventDefault();
    router.push(`/cart`);
    
    await addToCart();
  }
  
  return (
    <form onSubmit={handleAddToCart}>
      <button type="submit>Submit</button>
    </form>
  )
}
```

**Comment by lukepearce** [comment]
Further discussion can be read on [Discord](https://discord.com/channels/752553802359505017/1157302253955321957)

**Comment by lukepearce** [comment]
I found a workaround for the time being. If I add an inline function, the `router.push()` is executed before the server action starts to run.

```test.tsx
<form
  onSubmit={(e) => {
    e.preventDefault();
    router.push(`/${locale}/cart`);
    handleAddToCart;
  }}
  className="mb-6"
>
```

**Comment by amitbeck** [comment]
I'm experiencing the same issue. My workaround was to pass the path to the server action and perform a [redirect](https://nextjs.org/docs/app/api-reference/functions/redirect) in the server action. Not ideal but works.
